feat: add dynamic array encoding for encodeData

fhildeb opened the pull request:
### What kind of change does this PR introduce?

**Feature**: Dynamically encode ERC725 storage keys of type **Array**.

### What is the current behavior?

The library only allows users to define the full array they want to encode and set to a contract. For updates, developers must read through all existing elements and pay the gas for re-setting elements in the list, even if only a few of them are actually updated. This is highly impractical. 

### What is the new behavior?

Within the **encodeData(...)** function, developers can define a **startingIndex** and an **arrayLenght** parameter to dynamically adjust the output, which can be used for more gradient storage updates. Both parameters are optional. If not provided, the encoding falls back to the default behavior as before, starting from index 0 and using the length of the provided elements in the array.

Hugoo reviewed (changes requested):
Can you also update the function JSDoc at lines 231, 232. It is missing the 2 new inputs.

Can you update the docs and add examples on this new feature?

Nice tests

Hugoo commented on src/lib/utils.test.ts:
```diff
@@ -720,6 +720,168 @@ describe('utils', () => {
+              '0x983abc616f2442bab7a917e6bb8660df8b01f3bf',
+              '0x56ecbc104136d00eb37aa0dce60e075f10292d81',
+            ],
+            arrayLength: 23,
```
I have the feeling this can be a bit flaky. Like ppl can mess up a bit with this.

What about changing this to "currentArrayLength"

And then, the value you return is `currentArrayLength + value.length`

Then there is no way to have a wrong number, or at least it is harder to fail it.


Because right now, if let's say:

- current array of size = 5
- you add 3 values
- you give `arrayLength=4`

You have a valid (as per your code) values.length > arrayLength. But you have a wrong value of array length = 4 but it should be 5 + 3 = 8.

I think this suggestion might make it less error prone.

what do you think?

Hugoo replied:
For next time, I suggest to open issues and have the discussion on the issue page so everyone (incl. me and external Devs) can get the whole discussion 😉

Hugoo replied:
Then maybe you can call it "finalArrayLength"?

Or something. Just as arrayLength can be confusing. Is it before? After? etc.

Hugoo commented on src/lib/utils.ts:
```diff
@@ -260,21 +262,46 @@ export function encodeKey(
         return null;
       }
 
+      // Check if an starting index and array lenght are correct types
```
Actually, you can remove this comment. It is so obvious from the code. This comment is not needed.

Less code less problems.

Hugoo commented on docs/classes/ERC725.md:
```diff
@@ -700,6 +702,50 @@ myErc725.encodeData([
+  },
+];
+
+const myErc725 = new ERC725(schemas);
```
Please keep this example consistent with the other examples of that function.

1. No instantiation of ERC725 (why would this one example have `new ERC725(...)` and not the other ones? Either all get it or none.
2. You can use it like that:

```js
const schemas = [
  {
    name: 'DynamicKey:<bytes4>:<string>',
    key: '0xForDynamicKeysThisFieldIsIrrelevantAndWillBeOverwriten',
    keyType: 'Mapping',
    valueType: 'bytes',
    valueContent: 'Address',
  },
];

myErc725.encodeData(
  [
    {
      keyName: 'DynamicKey:<bytes4>:<string>',
      dynamicKeyParts: ['0x11223344', 'Summer'],
      value: '0xcafecafecafecafecafecafecafecafecafecafe',
    },
  ],
  schemas,
);
```

fhildeb commented:
Thanks for the review and feedback 🙏🏻

I renamed `arrayLength` to `totalArrayLength` for better clarification. I also added a new info box and warning regarding the usage of these elements. 

I think it's important to eliminate all questions, as this will be the second "big update" for dynamic user inputs, besides the already existing `dynamicKeyParts`...which also has some extended explanation in the docs. 👍🏻

CJ42 commented on src/index.test.ts:
```diff
@@ -1004,6 +1004,45 @@ describe('Running @erc725/erc725.js tests...', () => {
           assert.deepStrictEqual(results, intendedResult);
         });
 
+        it('encodes dynamic data values for keyType "Array" in naked class instance', () => {
```
name of test seems incorrect.

CJ42 commented on src/lib/utils.ts:
```diff
@@ -260,21 +262,45 @@ export function encodeKey(
+
+      if (startingIndex < 0) {
+        throw new Error(
+          'Invalid startingIndex parameter. Value cannot be negative.',
```
Suggest showing variable in quote in error string.

```suggestion
          'Invalid `startingIndex` parameter. Value cannot be negative.',
```

fhildeb replied:
I was not able to add brackets or double quotes in Error outputs. Tried both `"`, ```, `\`` due to the linter configuration.

@CJ42 do you have another idea to improve the outputs?

CJ42 replied:
Maybe if you escape them with backslash `\`. But it is a minor detail, it is fine as it is.

fhildeb replied:
As you see above, I tried backslash, but the linter removes it automatically. I don't know what's better- ignore formatting and linting for those lines or keep it as is. I would argue the second one is more convenient.

When I edit the error messages, the `throw` tests also start to fail:

<img width="911" alt="img1" src="https://github.com/ERC725Alliance/erc725.js/assets/61689369/20c0d327-6072-41ad-9816-8daa1bc6c296">

richtera replied:
You don't need a backslash for different quotes. What was the error when you put a plain backquote around the startingIndex and totalArrayLength? Usually error message don't contain back quotes, but I don't really see a reason why not.

CJ42 commented on src/lib/utils.ts:
```diff
@@ -260,21 +262,45 @@ export function encodeKey(
+
+      if (totalArrayLength < value.length) {
+        throw new Error(
+          'Invalid totalArrayLength parameter. Array length must be at least as large as the number of elements of the value array.',
```
same comment as above

CJ42 reviewed (approved):
LGTM

fhildeb commented:
Alright, I figured out how to adjust the error messages. Had to adjust the tests to also include these brackets, as they were looking for a close match.
